### DISCUSSION TITLE
Support for arm64?

### DISCUSSION THREAD

**Discussion by htrendafilov** [super-linter/super-linter#1956] [Q&A]
Are there any plans for this image to be multi-arch? /ARM support/

Sometimes it's convenient to be able to run it locally and with new Macs with ARM processor I couldn't do it.

**Comment by BretFisher** [comment]
+1 on the idea. I'm a arm64/v8 and arm/v7 fan. I think the trick there is that *every* linter would have to support arm64 or be custom-built during an arm64 container image build. That's no small effort IMO.  I have a [toolkit image](https://github.com/BretFisher/shpod) that I build cross-platform ([here's what the Dockerifle looks like to deal w/ multiplatform](https://github.com/BretFisher/shpod/blob/6ac6db5baaff998821c22ee90bf9d53eb9990551/Dockerfile)) and it was some serious work. Super-Linter is many more tools then that. ([here's some notes I took on that process](https://github.com/BretFisher/multi-platform-docker-build))

@htrendafilov I'm on a M1 Pro and it works fine on Docker Desktop because [QEMU](https://www.qemu.org/) emulation is built in, so a `docker run -e RUN_LOCAL=true -v /path/to/local/codebase:/tmp/lint github/super-linter` pulls the x86_64 image and then docker runs it via emulation, which a lot of docker images uses because they don't have an arm64 image. It's meant to be transparent to the user.

If you're using something other than Docker Desktop, you can google how to enable QEMU for that Linux distro.

**Comment by BretFisher** [comment]
I'm adding some more details on an attempt I made today to build the Dockerfile on arm64, for when someone wants to attempt a PR.

This is a much bigger PR from what I can tell (which is what I was saying in https://github.com/github/super-linter/discussions/1956#discussioncomment-2347889).  The current build assumes x86_64. An arm64 build shouldn't assume the docker build server or the Actions runner has QEMU/Rosetta enabled (to emulate x86_64 on arm64). *Every* tool used or built needs to be for arm64.

### Things that would need to be included in an arm64 PR:

- **Change the Dockerfile install style for many of the tools.** The Dockerfile pulls lots of binary releases via the easy "copy a file from another docker image" but many of those images don't have an arm64 variant. For example, the [alpine/terragrunt](https://hub.docker.com/r/alpine/terragrunt/tags) image, which has had a [request for over a year](https://github.com/alpine-docker/terragrunt/issues/21) to build an arm64 image. In this case of terragrunt, we'd need to move to curling [upstream binaries](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.45.2), which **does** have arm64. This is an industry-wide issue that some of us have been working on, but it requires PRs (and often troubleshooting) across a wide range of repos. We'll likely find some tools don't even have a linux/arm64 build yet.
- **Support runtime architecture downloads in scripts.** The installers in `/scripts` are also assuming x86_64, with some being hardcoded and others not dynamically choosing the arch when downloading.
- **Update CI and CD for multi-platform.** Adding `docker/setup-buildx-action` and `platforms: linux/amd64,linux/arm64` to docker builds in both `ci.yml` and `cd.yml`
- **Deal with the lack of arm64 public runners.** Not everything is supported in QEMU, which is currently the only way to build an arm64 image on public runners. We may run into build fails due to a QEMU edge case. If any potential QEMU issues can't be resolved, the options are then building on a private runner (AWS Graviton) or just holding until GitHub supports arm64 public runners (which [aren't yet planned](https://github.com/actions/runner/issues/805#issuecomment-1438193387) or [on the roadmap](https://github.com/github/roadmap/issues?q=arm64).)

**Comment by denisa** [comment]
For reference, here is where the docker images stand on =/21/23:

# With arm:

- alpine/terragrunt:1.4.5
- assignuser/chktex-alpine:v0.2.0
- dotenvlinter/dotenv-linter:3.3.0
- golangci/golangci-lint:v1.52.2
- ghcr.io/terraform-linters/tflint-bundle:v0.46.0.1
- ghcr.io/yannh/kubeconform:v0.6.2
- hadolint/hadolint:latest-alpine
- hashicorp/terraform:1.4.4
- koalaman/shellcheck:v0.9.0
- mstruebing/editorconfig-checker:2.4.0
- mvdan/shfmt:v3.6.0
- rhysd/actionlint:1.6.24
- yoheimuta/protolint:0.43.1
- zricethezav/gitleaks:v8.16.2
- python:3.11.1-alpine3.17
- tenable/terrascan:? (support on main branch, in release post 1.18.3)

# Without arm:

- ~alpine/terragrunt:1.4.4~
- ~assignuser/chktex-alpine:v0.1.1~
- cljkondo/clj-kondo:2023.03.17-alpine ([issue](https://github.com/clj-kondo/clj-kondo/issues/2093))
- ghcr.io/awkbar-devops/clang-format:v1.0.2 ([issue](https://github.com/Awkbar-DevOps/clang-format/issues/62))
- ~ghcr.io/yannh/kubeconform:v0.6.1~
- scalameta/scalafmt:v3.7.3 ([issue](https://github.com/scalameta/scalafmt/issues/3542))
- ~tenable/terrascan:1.17.1~

**Reply by BretFisher** [reply]
So I'm thinking "without Arm" means we need to either:

1. lobby/PR the image-building repo to add arm64 builds
2. or just pull upstream binary rather than docker image copy
3. or lobby/PR upstream repo to build a binary

WDYT?

**Reply by denisa** [reply]
not sure yet…

I believe a few of these images would be easy to improve, but it depends on the readiness of the owner, doesn’t it?

I want to check what the lift is to, at least, update super-linter:slim image to become multi-arch, with the hope that this would incentivize the remaining dependencies

**Reply by denisa** [reply]
assignuser/chktex-alpine hasn’t been touched for 3 years, and its https://github.com/assignUser/chktex-alpine/blob/master/Dockerfile is very simple, I think that would be a good case for creating a bash install

**Comment by denisa** [comment]
As for the scripts, I parametrized them on a branch in my fork.

# probably ready to go

- install-dotnet.sh
   (dotnet installer supports amd64 and arm64)
- install-lintr.sh
   (r is a scripting system available on amd64 and arm64)
- install-pwsh.sh
   (linux binaries are available on the release channel for amd64 and arm64)
- install-raku.sh
   (lua «is available every a c compiler is available» and there are multi-arch docker images)

# probably not compatible

- install-ktlint.sh
   (expects a glibc for alpine)
- install-phive.sh
   (expects a glibc for alpine)

# to be investigated

- install-arm-ttk.sh

**Reply by BretFisher** [reply]
Note the image [already has glibc](https://github.com/github/super-linter/blob/47884a0e25aabccfc81d464b57cb5c8af37af6d8/Dockerfile#L43) it seems

**Comment by denisa** [comment]
Note sure if this is helpful, I tried to summarize in a table the arm64 compatibility (2nd column) and the images in which it is used (3rd and 4th columns)

The intention being to see if prioritizing the slim image helps define an MVP

 dependency | arm64? | slim | full 
------------|--------|------|------
dotenvlinter/dotenv-linter:3.3.0 | y | n | y
golangci/golangci-lint:v1.52.2 | y | y | y
ghcr.io/terraform-linters/tflint-bundle:v0.46.0.1 | y | y | y
hadolint/hadolint:latest-alpine | y | y | y
hashicorp/terraform:1.4.4 | y | y | y
koalaman/shellcheck:v0.9.0 | y | y | y
mstruebing/editorconfig-checker:2.4.0 | y | y | y
mvdan/shfmt:v3.6.0 | y | y | y
rhysd/actionlint:1.6.24 | y | y | y
yoheimuta/protolint:0.43.1 | y | y | y
zricethezav/gitleaks:v8.16.2 | y | y | y
python:3.11.1-alpine3.17 | y | y | y
alpine/terragrunt:1.4.4 | y | y | y | as of 1.4.5
assignuser/chktex-alpine:v0.2.0 | y | y | y
**cljkondo/clj-kondo:2023.03.17-alpine** | n | y | y
**ghcr.io/awkbar-devops/clang-format:v1.0.2** | n | y | y
ghcr.io/yannh/kubeconform:v0.6.1 | y | y | y
**scalameta/scalafmt:v3.7.3** | n | y | y
**tenable/terrascan:1.17.1** | n | y | y
install-dotnet.sh | y? | n | y
install-lintr.sh | y? | y | y
install-pwsh.sh | ubuntu only | n | y 
install-raku.sh | y? | y | y
install-rustfmt.sh | y? | n | y
**install-ktlint.sh** | n? | y | y
**install-phive.sh** | n? | y | y
install-arm-ttk.sh | ? | n | y

**Comment by denisa** [comment]
terragrunt is now multi arch (https://github.com/alpine-docker/terragrunt/pull/25)

**Comment by ferrarimarco** [comment]
Can you please open a new feature request for this, so we can track this workstream? Thanks!
